token 2022: Expose token-metadata state decoder in @solana/token-metata

mistersimon opened the pull request:
Token2022 doesn't using 8-byte discriminators, and can't use the unpack function current in @solana/spl-token-metadata which assumes 8-byte discriminators.

This PR exposes a function that just decodes to the data, so it can be used in Token2022.

Discussion where the need arose:
https://github.com/solana-labs/solana-program-library/pull/5667#discussion_r1390234922

buffalojoec reviewed (changes requested):
Ok, so here's what we want to do (and what's in Rust):

In `@solana/spl-token-metadata`: Provide an `unpack` function that _just_ uses the deserialization to unpack a buffer into `TokenMetadata`; no discriminator _or_ length

In `@solana/spl-token`: Use that `unpack` in the token metadata extension to deserialize the token metadata from the extension. The discriminator and length bits are taken care of for you when you call `getExtensionData`, so we just need a function that calls that for the `TokenMetadata` extension, and then unpacks the returned buffer into `TokenMetadata`!

Here's an example:

https://github.com/solana-labs/solana-program-library/blob/f542e00a8905eaf3f55dea1c08ca4787138e1788/token/js/src/extensions/transferFee/state.ts#L92-L99

mistersimon commented:
> Ok, so here's what we want to do (and what's in Rust):
> 
> In `@solana/spl-token-metadata`: Provide an `unpack` function that _just_ uses the deserialization to unpack a buffer into `TokenMetadata`; no discriminator _or_ length


If i'm understanding correctly, the `unpack` function you describing is the function `decodeTokenMetadata` I added.
https://github.com/solana-labs/solana-program-library/pull/5806#pullrequestreview-1726249136

I left the behaviour of the `unpack` function the same to avoid a breaking change, but it uses `decodeTokenMetadata` under the hood. 

You suggesting we get rid of the `unpack` function as its implemented now and just leave `decodeTokenMetadata` but renamed to `unpack`?

buffalojoec commented:
> You suggesting we get rid of the `unpack` function as its implemented now and just leave `decodeTokenMetadata` but renamed to `unpack`?

Yes, we're basically just copying this, which is serializing/deserializing with `borsh`.

https://github.com/solana-labs/solana-program-library/blob/b1ba6493fb49a89fcbb0abcf0b5f05694a8bf7d1/token-metadata/interface/src/state.rs#L99-L109

We could even flop the `decoder` object from `@solana/codecs-data-structures` to use `codec` from and then we can roll `pack` and `unpack`! Although I don't know why we would need pack in JS at the moment.

Regardless, I think we should opt for `unpack` and `pack` naming conventions over `decode`, etc.

mistersimon commented on token-metadata/js/package.json:
```diff
@@ -1,7 +1,7 @@
 {
     "name": "@solana/spl-token-metadata",
     "description": "SPL Token Metadata Interface JS API",
-    "version": "0.1.0",
```
How do you all manage versions? Do you strictly follow semver? Cause then i think it should be 1.0.0

buffalojoec replied:
I'll do the version bump separately!

mistersimon commented on token-metadata/js/src/state.ts:
```diff
@@ -1,9 +1,15 @@
+} from '@solana/codecs-data-structures';
+import { getStringDecoder, getStringEncoder } from '@solana/codecs-strings';
 
 export const TOKEN_METADATA_DISCRIMINATOR = Buffer.from([112, 132, 90, 90, 11, 88, 157, 87]);
```
I think it still makes sense to expose this even though it is not used, since i understand its a standardised value?

buffalojoec replied:
Yeah sure.

mistersimon commented:
> Regardless, I think we should opt for `unpack` and `pack` naming conventions over `decode`, etc.

Okay i changed the behaviour of unpack to what you described. 

Also added pack, as you say not sure there is a need but it was already there as a helper for the test cases - just had to move and expose it.

buffalojoec commented on token-metadata/js/test/state.test.ts:
```diff
@@ -2,68 +2,31 @@ import { PublicKey } from '@solana/web3.js';
-        const meta = unpack(input);
-        expect(meta).to.deep.equal({
+    it('Can pack and unpack as rust implementation', () => {
+        const meta = {
```
I just noticed this test doesn't use `updateAuthority` in the data structure. Can you add to the test a step where you add `updateAuthority` and test the bytes again? It doesn't have to be a separate test.

mistersimon replied:
I modified the last test which has the updateAuthority to check against the bytes. 

https://github.com/solana-labs/solana-program-library/pull/5806/commits/3839116439f00552319248f01d2af4206ae954ae#diff-3223c36003634dbc6a302f5e66b19f6aed12d8aa50040ebf6b79ec4fa3a45c5fR52

buffalojoec commented on token-metadata/js/package.json:
```diff
@@ -1,7 +1,7 @@
     "name": "@solana/spl-token-metadata",
     "description": "SPL Token Metadata Interface JS API",
-    "version": "0.1.0",
+    "version": "0.2.0",
```
As I mentioned I'll bump this!

mistersimon replied:
Sure. Reverted back to 0.1.0 and i'll leave you to bump

buffalojoec commented on token-metadata/js/src/state.ts:
```diff
@@ -32,12 +38,28 @@ function isNonePubkey(buffer: Uint8Array): boolean {
-    }
+// Pack TokenMetadata into byte slab
+export const pack = (meta: TokenMetadata): Uint8Array => {
+    const encoder = getStructEncoder([
+        ['updateAuthority', getBytesEncoder({ size: 32 })],
+        ['mint', getBytesEncoder({ size: 32 })],
+        ['name', getStringEncoder()],
+        ['symbol', getStringEncoder()],
+        ['uri', getStringEncoder()],
+        ['additionalMetadata', getArrayEncoder(getTupleEncoder([getStringEncoder(), getStringEncoder()]))],
+    ]);
```
You can save the code duplication here by instead importing `getArrayCodec`, `getBytesCodec`, etc. and define this as `codec`. Then it will yield both `encode` and `decode` from the same constant object.

mistersimon replied:
Oh i didn't know there was that functionality. That will be a lot cleaner. Changed it 👍

buffalojoec reviewed (commented):
Conceptually this is perfect, let's just tighten up a few bits.

mistersimon commented:
> Conceptually this is perfect, let's just tighten up a few bits.

Thanks for the review - I've pick up those pieces. Let me know if there is anything else, otherwise will need to trouble you to merge + publish to npm. Thanks in advance!

buffalojoec commented on token-metadata/js/src/state.ts:
```diff
@@ -1,18 +1,18 @@
 
 export const TOKEN_METADATA_DISCRIMINATOR = Buffer.from([112, 132, 90, 90, 11, 88, 157, 87]);
 
+export const tokenMetadataCodec = getStructCodec([
```
Let's _not_ export this, since we are peppering in some custom logic for all-zero update authorities, etc. Someone might get confused if they try to use the codec in its raw form.

buffalojoec commented on token-metadata/js/test/state.test.ts:
```diff
@@ -50,23 +50,23 @@ describe('Token Metadata State', () => {
 
     it('Can pack and unpack with mint and updateAuthority', () => {
-        const input = pack({
+        const meta = {
```
We should also probably add a test with some additional metadata.

buffalojoec reviewed (commented):
Thanks! Couple small things then we're good to go.

buffalojoec commented on token-metadata/js/src/state.ts:
```diff
@@ -4,7 +4,8 @@ import { getStringCodec } from '@solana/codecs-strings';
 export const TOKEN_METADATA_DISCRIMINATOR = Buffer.from([112, 132, 90, 90, 11, 88, 157, 87]);
 
-export const tokenMetadataCodec = getStructCodec([
+// Don't export as you still need custom logic for all-zero update authorities
```
Not necessary, lol.

buffalojoec reviewed (approved):
![nice](https://media.giphy.com/media/mCIjCgs3nWQWfJZvPA/giphy-downsized-large.gif)
